torch/_inductor/fx_passes/post_grad.py

```diff
@@ -484,6 +576,125 @@ def is_valid_splitwithsizes_cat(match):
+        if start is None or end is None:
+            return False
+        if start == 0 and end >= 2**63 - 1 and step == 1:
+            return True
+        return False
+
+    def slice_cond(self, dim=0, start=None, end=None, step=1):
+        if start is None or end is None:
+            return False
+        if start == 0 and end >= 2**63 - 1 and step == 1:
+            return True
+        return False
+
+    input_storages = set()
+    output_storages = set()
+    for node in graph.nodes:
+        if node.op == "placeholder":
+            input_storages.add(get_node_storage(node))
+        else:
+            break
+
+    for out in tuple(graph.nodes)[-1].args[0]:
+        if isinstance(out, torch.fx.Node):
+            output_storages.add(get_node_storage(out))
+
+    noop_conditions = {
+        aten.slice.Tensor: (0, slice_cond),
+        aten.clone.default: (0, true),
+        aten.slice_scatter.default: (1, slice_scatter_cond),
+    }
+
+    for node in graph.nodes:
+        if node.target in noop_conditions:
+            cloned_arg_index, cond = noop_conditions[node.target]
+            src = node.args[cloned_arg_index]
+            if (
+                get_node_storage(node) in output_storages
+                and get_node_storage(src) in input_storages
+            ):
+                continue
+            if same_layout(node, src) and cond(*node.args, **node.kwargs):
+                node.replace_all_uses_with(src)
+                graph.erase_node(node)
+
+
+def reinplace_scatters(graph):
+    """
+    Reinplaces scatter operations in easy cases where the node being mutated
+    is only used by the scatter (users == 1), and the node being mutated
+    shares storage with no other nodes.
+
+    Also handles input mutations when there is a corresponding copy node.
+    """
+
+    copy_nodes = {}
+    mutated_inputs = set()
+    storage_to_nodes = defaultdict(list)
+    for node in reversed(graph.nodes):
+        if isinstance(node.meta["val"], torch.Tensor):
+            storage_to_nodes[get_node_storage(node)].append(node)
+        if node.target == aten.copy_.default:
+            copy_nodes[(node.args[0], node.args[1])] = node
+            mutated_inputs.add(node.args[0])
+        elif node.op == "output":
+            pass
+
+    InplaceableOp = namedtuple("InplaceableOp", ["inplace_op", "mutated_arg"])
+
+    inplaceable_ops = {
+        aten.index_put.default: InplaceableOp(aten.index_put_.default, 0),
+    }
+    for node in graph.nodes:
+        if inplaceable_op := inplaceable_ops.get(node.target, False):
+            mutated_arg = node.args[inplaceable_op.mutated_arg]
+            if get_node_storage(mutated_arg) is None:
+                continue
+            shared_view_nodes = storage_to_nodes[get_node_storage(mutated_arg)]
+            if mutated_arg.op == "placeholder":
+                if not (copy_node := copy_nodes.get((mutated_arg, node), False)):
+                    continue
```

Comment:
i guess it's an invariant that a node storage will only ever have a single copy

Reply:
Note, @bdhirsh is working on a change that would break that invariant. I dont thinkkkk that it affects this pass in terms of accuracy, though. But it will make using storages as aliasing less accurate. 

Im not sure what exactly is the most clean fix, maybe forcing the FakeTensor conversion to create separate Storages for definitely not aliasing/overlapping inputs.